libs/x-content/src/main/java/org/elasticsearch/common/xcontent/InstantiatingObjectParser.java

```diff
@@ -0,0 +1,191 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.ParseField;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Like {@link ConstructingObjectParser} but works with objects which have a constructor that matches declared fields.
+ * <p>
+ * Declaring a {@linkplain InstantiatingObjectParser} is intentionally quite similar to declaring an {@linkplain ConstructingObjectParser}
+ * with two important differences.
+ * <p>
+ * The first differences being that constructor arguments are declared with the consumer returned by the static {@link #constructorArg()}
+ * method and that {@linkplain InstantiatingObjectParser}'s constructor takes a class of the target object. The target object must have
+ * exactly one constructor with the number and order of arguments matching the number of order of declared fields. If there are more
+ * then 2 constructors with the same number of arguments, one of them needs to be marked with {@linkplain ParserConstructor} annotation.
+ * </p>
+ * The second differences being that at the end of field declarations the {@linkplain #finalizeFields()} method has to be called before
+ * any parsing is possible.
+ * <pre>{@code
+ *   public static class Thing{
+ *       public Thing(String animal, String vegetable, int mineral) {
+ *           ....
+ *       }
+ *
+ *       public void setFruit(int fruit) { ... }
+ *
+ *       public void setBug(int bug) { ... }
+ *
+ *   }
+ *
+ *   private static final InstantiatingObjectParser<Thing, SomeContext> PARSER = new InstantiatingObjectParser<>("thing", Thing.class);
+ *   static {
+ *       PARSER.declareString(constructorArg(), new ParseField("animal"));
+ *       PARSER.declareString(constructorArg(), new ParseField("vegetable"));
+ *       PARSER.declareInt(optionalConstructorArg(), new ParseField("mineral"));
+ *       PARSER.declareInt(Thing::setFruit, new ParseField("fruit"));
+ *       PARSER.declareInt(Thing::setBug, new ParseField("bug"));
+ *       PARSER.finalizeFields()
+ *   }
+ * }</pre>
+ */
+public class InstantiatingObjectParser<Value, Context> extends AbstractConstructingObjectParser<Value, Context> {
+    private final Class<Value> valueClass;
+    private volatile Constructor<Value> constructor;
```

Comment:
I wonder if it'd make more sense to make a `Builder` for it rather than this. These are generally static state stuff like this is a bit scary. I see why it is just fine but it'd be a bit more "obviously fine" if it had a builder instead of "finalized".